### ISSUE TITLE
API v2

### ISSUE THREAD

**Issue by aurbano** [socialgorithm/ultimate-ttt-js#2]
A game of Ultimate TicTacToe is a board where each cell is a normal TicTacToe game. We represent this as two classes: Sub Board, and the Main Board.

The main difference is that one has a matrix of ints, and the other has a matrix of SubBoards.

They share a common API where applicable.

### Data structures
* `Cell: Object` Object representing a SubBoard cell.
  - `player: int` [`-1`: none, `0`: You, `1`: Opponent]
  - `index: int` index representing when this cell was played, starts at 0. It's `-1` if the cell hasn't been played yet.
* `board: Array<Array<Cell>>` represents a Tic Tac Toe board, it's a two dimensional array of `Cells`, described above.
* `move: Array<int>` that has `[row, column]` - this is the data structure expected when a method takes a move as a parameter.

### Methods
* `SubBoard(size=3): SubBoard` Constructor, returns a new board initialised to 0 in every cell
* `isFinished(): boolean` Returns true if the board is finished (tied, won, or lost)
* `getResult(): int` Returns the winner for the board, throws an exception if the game hasn't finished yet.
  - `-1` if it's a tie
  - `0` you won
  - `1` the opponent won
* `isValidMove(move): boolean` Returns true if the move is valid
* `addMyMove(move): SubBoard` Adds your move to the board, throws exception if move is invalid or board is already finished.
* `addOpponentMove(move): SubBoard` Adds an opponent move to the board, throws exception if move is invalid or board is already finished.
* `prettyPrint(): string` Returns a string with the board formatted for display

### Examples:

Testing a move
```js
board = new SubBoard()
print board.addMyMove([1, 2]).getResult()
// outputs 0
// board hasn't been modified

// let's assume I liked that move, store it:
board = board.addMyMove([1, 2])

print board.prettyPrint()

// would print:
// 0 0 0 
// 0 0 1
// 0 0 0
```

### Attributes
These attributes are considered "private" in the sense that you should interact with them via the API methods. But they are exposed "publicly" in case you know what you're doing and want to edit them.

* `board: Array<Array<SubBoard>>`, where the `SubBoard` is in the format defined above
* `stateBoard: SubBoard` a `SubBoard` object that represents the state of the `MainBoard`. Each cell contains the state of the `SubBoard.getResult()` method of the `SubBoard` it represents.
* `moves: int` Number of moves that have been played so far
* `lastMove: move` Coordinates of the last move

### Methods
* `MainBoard(size=3): MainBoard` Get a new `MainBoard` object, with all `SubBoards` initialised to 0's.
* `isFinished(): boolean` Returns true if the board is finished (tied, won, or lost)
* `getResult(): int` Returns the winner for the board
  - `-1` if the board isn't finished yet
  - `0` if it's a tie
  - `1` you won
  - `2` the opponent won
* `isValidBoardRowCol(boardRowCol): boolean` Returns true if the specified board `[row, col]` is a valid choice for the current game state taking into account the previous move and the state of the board.
* `isValidMove(boardRowCol, move): boolean` Returns true if the move is valid
* `addMyMove(boardRowCol, move): MainBoard` Adds your move to the board, throws exception if move is invalid or board is already finished.
* `addOpponentMove(boardRowCol, move): MainBoard` Adds an opponent move to the board, throws exception if move is invalid or board is already finished.
* `prettyPrint(): string` Returns a string with the board formatted for display

**Comment by aurbano** [comment]
@reddyalready Updated `SubBoard` docs to add a `Cell` definition instead of a simple int

**Comment by aurbano** [comment]
Implemented in #3
